src/evaluators/source-4-3.js

```diff
@@ -0,0 +1,911 @@
+}
+
+
+function declare_constant(name, val, env) {
+    set_name_value(name, val, env);
+}
+
+function analyze_variable_declaration(stmt) {
+    const name = variable_declaration_name(stmt);
+    const vfun = analyze(variable_declaration_value(stmt));
+    return (env, succeed, fail) =>
+        vfun(env,
+            (val, fail2) => {
+                declare_variable(name, val, env);
+                succeed("variable declaration ok", fail2);
+            },
+            fail);
+}
+
+function analyze_constant_declaration(stmt) {
+    const name =
+        constant_declaration_name(stmt);
+    const vfun =
+        analyze(constant_declaration_value(stmt));
+    return (env, succeed, fail) =>
+        vfun(env,
+            (val, fail2) => {
+                declare_constant(name, val, env);
+                succeed("constant declaration ok", fail2);
+            },
+            fail);
+}
+
+function analyze_assignment(stmt) {
+    const variable = assignment_name(stmt);
+    const vfun = analyze(assignment_value(stmt));
+    return (env, succeed, fail) =>
+        vfun(env,
+            (val, fail2) => { // *1*
+                const old_value = lookup_name_value(variable, env);
+                set_name_value(variable, val, env);
+                succeed("assignment ok",
+                    () => { // *2*
+                        set_name_value(variable, old_value, env);
+                        fail2();
+                    });
+            },
+            fail);
+}
+
+function analyze_application(stmt) {
+    const ffun = analyze(operator(stmt));
+    const afuns = map(analyze, operands(stmt));
+    return (env, succeed, fail) =>
+        ffun(env,
+            (fun, fail2) =>
+            get_args(afuns,
+                env,
+                (args, fail3) =>
+                execute_application(fun,
+                    args, succeed, fail3),
+                fail2),
+            fail);
+}
+
+function get_args(afuns, env, succeed, fail) {
+    return is_null(afuns) ?
+        succeed(null, fail) :
+        head(afuns)(env,
+            // success continuation for this afun
+            (arg, fail2) =>
+            get_args(tail(afuns),
+                env,
+                // success continuation for
+                // recursive call to get_args
+                (args, fail3) =>
+                succeed(pair(arg, args),
+                    fail3),
+                fail2),
+            fail);
+}
+
+function execute_application(fun, args, succeed, fail) {
+    if (is_primitive_function(fun)) {
+        succeed(apply_primitive_function(fun, args), fail);
+    } else if (is_compound_function(fun)) {
+        const body = function_body(fun);
+        const locals = function_locals(fun);
+        const names = insert_all(map(name_of_name, function_parameters(fun)), locals);
+        const temp_values = map(x => no_value_yet, locals);
+        const values = append(args, temp_values);
+
+        const application_value = function_body(fun)(
```

Comment:
Suggestion: instead of `function_body(fun)`, perhaps we can use `body` that's already declared in line 768?

Reply:
Thank you! Updated at 547be0b